### DISCUSSION TITLE
Why are VIFe-codes defined without the top-bit (the extension bit)?

### DISCUSSION THREAD

**Discussion by mfritz-gcx** [wmbusmeters/wmbusmeters#1369] [General]
Hey guys, 

I'm just diving into your code.
While doing so I wondered why the [extended VIF-codes](https://github.com/wmbusmeters/wmbusmeters/blob/aaf45d8bcd1954990507beed9b8ccd409ed36bd2/src/dvparser.h#L53) are defined with `0x7D` instead of `0xFD`? 

![image](https://github.com/user-attachments/assets/11167688-654b-40a4-bac0-c6cd10d82904)

What I understand from EN13757-3, section 6.4 is, that the top bit has to be set.

Does this has something to do with that the top bits are removed from the vif-byte, when they are exctracted ([wmbus.cc, Line 2656](https://github.com/wmbusmeters/wmbusmeters/blob/aaf45d8bcd1954990507beed9b8ccd409ed36bd2/src/wmbus.cc#L2656)).

Thanks for you help, I'm curious.

**Comment by weetmuts** [comment]
The top bits are great because theoretically you can find the number of difs and vifs without understanding the difs/vifs.
You only need to know how many value bytes there are according to the first dif byte. Then you can safely skip any dif/vif/value that you do not understand.

So, my idea was that I extract the dif/vifs separately and remove the top bits, since they have served their purpose.
For the the extension vifs, the top bit will be set, because it has to be... :-) because otherwise we will not include the next byte as the extended vif. As soon as you extracted the vifs, the top bits are no longer necessary. So, the lookup table does not use them.

Also the combinable vifs that appear at the end can occur both with a high bit or no hight bit depending if they are the last combinable or not....again, it makes sense to a single entry in the table (without top-bit) instead of two (with and without top bit).

Alas, there are of course exceptions and the variable size vif adds a requirement that you have to at least be able to detect this vif, read the vif length in the next byte to skip, and then continue vif decoding, where the bytes will have the top bit set.....

**Comment by weetmuts** [comment]
The simplest dif/vif is just two bytes without top-bits. As soon as the top bit is set for a dif or vif byte, then that means that the next byte should be included into the dif (or vif) that you are collecting. A dif or vif or both can be up to 10 bytes each. I think.

**Comment by mfritz-gcx** [comment]
So in a nutshell: 
As soon as the top bits aren't needed anymore they are discarded to make the code a little bit easier.
Do I summarize that right?

**Comment by weetmuts** [comment]
yes
